### ISSUE TITLE
Add implementation of Free

### ISSUE THREAD

**Issue by DamianReeves** [zio/zio-prelude#1168]
Add an implementation of Free to `zio-prelude`.

As per discussions with @jdegoes we want to call this something like: `ImperativeDSL`.

**Comment by sideeffffect** [comment]
You guys know that there is [ZPure](https://zio.dev/zio-prelude/zpure/) already, right?

**Comment by DamianReeves** [comment]
Yes we are aware, and we are attempting to use `ZPure` as a way of executing the Free grammar.
